src/util/expr.cpp

```diff
@@ -539,3 +539,134 @@ const_unique_depth_iteratort exprt::unique_depth_cbegin() const
 { return const_unique_depth_iteratort(*this); }
 const_unique_depth_iteratort exprt::unique_depth_cend() const
 { return const_unique_depth_iteratort(); }
+
+static bool precedence(const exprt &a, const exprt &b)
+{
+  // we use the precendences that most readers expect
+  // (i.e., the ones you learn in primary school),
+  // and stay clear of the surprising ones that C has
+  if((a.id()==ID_mult || a.id()==ID_div) &&
+     (b.id()==ID_plus || b.id()==ID_minus))
+    return true;
+
+  if((a.id()==ID_equal || a.id()==ID_notequal ||
+      a.id()==ID_lt || a.id()==ID_gt ||
+      a.id()==ID_le || a.id()==ID_ge) &&
+     (b.id()==ID_and || b.id()==ID_or))
+    return true;
+
+  return false;
+}
+
+// This turns a multi-ary expression into a string,
+// adding parentheses where indicated by the precedence.
+static std::string as_expr_multi_ary(const exprt &src)
+{
+  bool first=true;
+
+  std::string result;
+
+  for(const auto &op : src.operands())
+  {
+    if(first)
+      first=false;
+    else
+    {
+      result+=' ';
+      result+=id2string(src.id());
+      result+=' ';
+    }
+
+    bool need_parentheses=
+      !op.operands().empty() &&
+      !precedence(op, src);
+
+    if(need_parentheses)
+      result+='(';
+
+    result+=op.as_string();
+
+    if(need_parentheses)
+      result+=')';
+  }
+
+  return result;
+}
+
+static std::string as_expr(const constant_exprt &src)
+{
+  auto type=src.type().id();
+
+  if(type==ID_bool)
+  {
+    if(src.is_true())
+      return "true";
+    else if(src.is_false())
+      return "false";
+    else
+      return src.pretty();
+  }
+  else if(type==ID_unsignedbv || type==ID_signedbv)
+  {
+    return integer2string(*numeric_cast<mp_integer>(src));
+  }
+  else if(type==ID_integer)
+  {
+    return id2string(src.get_value());
+  }
+  else if(type==ID_floatbv)
+  {
+    auto f=ieee_floatt(src);
+    return f.to_ansi_c_string();
+  }
+  else
+    return src.pretty();
+}
+
+// The below generates a string in a generic syntax
+// that is inspired by C/C++/Java, and is meant for debugging
+// purposes.
+std::string exprt::as_string() const
+{
+  if(id()==ID_plus || id()==ID_minus || id()==ID_mult ||
+     id()==ID_and || id()==ID_or)
+    return as_expr_multi_ary(*this);
+  else if(id()==ID_not)
+    return "!("+to_not_expr(*this).op().as_string()+')';
+  else if(id()==ID_constant)
+    return as_expr(to_constant_expr(*this));
+  else if(id()==ID_typecast)
+    return "cast("+to_typecast_expr(*this).op().as_string()+", "+
+           type().as_string()+')';
+  else if(id()==ID_if)
+  {
+    auto &if_expr=to_if_expr(*this);
+    return if_expr.cond().as_string()+'?'+
+           if_expr.true_case().as_string()+':'+
+           if_expr.false_case().as_string();
+  }
+  else
+  {
+    if(has_operands())
+    {
```

Comment:
- Unnecessary `else` after return.
 - `if(!has_operands()) return id_string()` to leverage early-return and de-indent the block below.